swap/src/bitcoin/wallet.rs

```diff
@@ -353,55 +350,69 @@ fn estimate_fee(
     transfer_amount: Amount,
     fee_rate: FeeRate,
     min_relay_fee: Amount,
-) -> Amount {
-    // Doing some heavy math here :)
-    // `usize` is 32 or 64 bits wide, but `f32`'s mantissa is only 23 bits wide.
-    // This is fine because such a big transaction cannot exist and there are also
-    // no negative fees.
-    #[allow(
-        clippy::cast_precision_loss,
-        clippy::cast_possible_truncation,
-        clippy::cast_sign_loss
-    )]
-    let sats_per_vbyte = ((weight as f32) / 4.0 * fee_rate.as_sat_vb()) as u64;
+) -> Result<Amount> {
+    if transfer_amount.as_sat() <= 546 {
+        bail!("Amounts needs to be greater than Bitcoin dust amount.")
+    }
+    if fee_rate.as_sat_vb() <= 0.0 {
+        bail!("Fee rate needs to be > 0")
+    }
+    if fee_rate.as_sat_vb() >= 100_000_000.0 || min_relay_fee.as_sat() > 100_000_000 {
+        bail!("A fee_rate or min_relay_fee of > 1BTC does not make sense")
+    }
+
+    let min_relay_fee = if min_relay_fee.as_sat() == 0 {
+        // if min_relay_fee is 0 we don't fail, we just set it to 1 satoshi;
+        Amount::ONE_SAT
+    } else {
+        min_relay_fee
+    };
+
+    let weight = Decimal::from(weight);
+    let weight_factor = Decimal::from_f32(4.0).context("Could not parse 0.4 :D")?;
+    let fee_rate = Decimal::from_f32(fee_rate.as_sat_vb()).context("Could not parse fee_rate.")?;
```

Comment:
Might be worth changing `as_sat_vb` to return you a `Decimal`?

Reply:
Do you mean changing `fee_rate.as_sat_vb()` to return a Decimal? That's not in our control, it's coming from `bdk`.